hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/update/server/SCMCRLUpdateHandler.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.hadoop.hdds.security.x509.crl.CRLCodec;
+import org.apache.hadoop.hdds.security.x509.crl.CRLInfo;
+import org.apache.ratis.thirdparty.io.grpc.Status;
+import org.apache.ratis.thirdparty.io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Class handle the CRL client update and response.
+ */
+public class SCMCRLUpdateHandler implements SCMUpdateHandler {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SCMCRLUpdateHandler.class);
+  private final CRLStore crlStore;
+
+  private static final SCMUpdateServiceProtos.Type TYPE =
+      SCMUpdateServiceProtos.Type.CRLUpdate;
+
+  private final Map<UUID, CRLClientInfo> clients;
+
+  SCMCRLUpdateHandler(CRLStore crlStore) {
+    this.crlStore = crlStore;
+    clients = new ConcurrentHashMap<>();
+  }
+
+  public SCMUpdateServiceProtos.Type getType() {
+    return TYPE;
+  }
+
+  @Override
+  public void handleClientRequest(SCMUpdateServiceProtos.UpdateRequest request,
+      SCMUpdateClientInfo clientInfo) {
+    SCMUpdateServiceProtos.CRLUpdateRequest updateStatusRequest =
+        request.getCrlUpdateRequest();
+    long clientCrlId = updateStatusRequest.getReceivedCrlId();
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Client {} updateStatus \nclientCrlId {}  \npendingCrls {}",
+          clientInfo.getClientId(), clientCrlId,
+          updateStatusRequest.getPendingCrlIdsList().toString());
+    }
+
+    CRLClientInfo crlClientInfo;
+    if (!clients.containsKey(clientInfo.getClientId())) {
+      crlClientInfo = new CRLClientInfo(clientInfo);
+      clients.put(clientInfo.getClientId(), crlClientInfo);
+    } else {
+      crlClientInfo = clients.get(clientInfo.getClientId());
+    }
+
+    crlClientInfo.setPendingCrlIds(
+        request.getCrlUpdateRequest().getPendingCrlIdsList());
+    crlClientInfo.setReceivedCrlId(
+        request.getCrlUpdateRequest().getReceivedCrlId());
+
+    sendCrlUpdateToClient(crlClientInfo);
+  }
+
+  @Override
+  public void onUpdate() {
+    LOG.debug("Update due to certificate revoke");
+    // server crl id is usually > client crl id when this is invoked.
+    clients.values().forEach(client -> {
+      sendCrlUpdateToClient(client);
+    });
+  }
+
+  @Override
+  public void onRemoveClient(SCMUpdateClientInfo clientInfo) {
+    clients.remove(clientInfo.getClientId());
+  }
+
+  private void sendCrlUpdateToClient(CRLClientInfo client) {
+    long clientCrlId = client.getReceivedCrlId();
+    long serverCrlId = crlStore.getLatestCrlId();
+
+    if (clientCrlId >= serverCrlId) {
+      return;
+    }
+
+    LOG.debug("## Server: clientCrlId {} serverCrlId {}",
+        clientCrlId, serverCrlId);
+
+    long nextCrlId = clientCrlId + 1;
+    try {
+      CRLInfo crlInfo = null;
+      while (crlInfo == null && nextCrlId <= serverCrlId) {
```

Comment:
Why do we need the null check here? This can be replaced with if
```suggestion
      if (nextCrlId <= serverCrlId) {
```

Reply:
This is to handle potential holes in the crlId sequence (might come from unit test). Server can still send the next crlId to client even though there is a gap. Correctness wise, I think we should throw an exception here if server has a gap in the crl sequence id.